### PR TITLE
iOS wrapper for alohalytics.

### PR DESCRIPTION
**Pull request by deathbaba**

### CODE REVIEW

File: examples/ios/SampleClient/AppDelegate.m [hunk 1]
```diff
@@ -33,29 +35,50 @@ @implementation AppDelegate
...
   // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
+  [Alohalytics logEvent:@"$didBecomeActive"];
 }
 
 - (void)applicationWillTerminate:(UIApplication *)application {
   // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
+  [Alohalytics logEvent:@"$willTerminate"];
+}
+
+// You need to set "fetch" in UIBackgroundsModes in your plist.
```
Comment (dkorolev) [inline examples/ios/SampleClient/AppDelegate.m]: Don't understand this comment.

File: examples/ios/SampleClient/AppDelegate.m [hunk 2]
```diff
@@ -33,29 +35,50 @@ @implementation AppDelegate
...
+-(void)application:(UIApplication *)application performFetchWithCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
+{
+  [Alohalytics forceUpload];
```
Comment (dkorolev) [inline examples/ios/SampleClient/AppDelegate.m]: Reminder: we spoke of `forceUpload` being the wrong call to make if you need the exponential delays to be respected.

File: src/apple/alohalytics_objc.mm [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  return map;
+}
+
+// Safe extraction from [possible nil] CLLocation to alohalytics::Location.
+static Location ExtractLocation(CLLocation * l) {
+  Location extracted;
+  if (l) {
+    if (l.horizontalAccuracy >= 0) {
+      extracted.SetLatLon([l.timestamp timeIntervalSince1970] * 1000.,
+                           l.coordinate.latitude, l.coordinate.longitude,
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: I believe this indent is off by one space.

File: src/apple/alohalytics_objc.mm [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                           l.horizontalAccuracy);
+    }
+    if (l.verticalAccuracy >= 0) {
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: These conditions might benefit from a comment that `accuracy < 0 <=> location is unknown`.

File: src/apple/alohalytics_objc.mm [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+  NSDictionary * attributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[[NSBundle mainBundle] executablePath] error:nil];
+  if (attributes) {
+    NSDate * date = [attributes objectForKey:NSFileModificationDate];
+    return std::to_string(static_cast<uint64_t>([date timeIntervalSince1970] * 1000.));
+  }
+  return std::string("0");
+}
+
+static std::string RectToString(CGRect const & rect) {
+  return std::to_string(rect.origin.x) + " " + std::to_string(rect.origin.y) + " "
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: High-level comment: You seem to do a lot of string packing here. Which brings us back to the question: Why not JSON?

File: src/apple/alohalytics_objc.mm [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+  + std::to_string(rect.size.width) + " " + std::to_string(rect.size.height);
+}
+
+// Logs some basic device's info.
+static void LogSystemInformation() {
+  UIDevice * device = [UIDevice currentDevice];
+  UIScreen * screen = [UIScreen mainScreen];
+  std::string preferredLanguages;
+  for (NSString * lang in [NSLocale preferredLanguages]) {
+    preferredLanguages += [lang UTF8String] + std::string(" ");
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: High-level comment: You seem to do a lot of string packing here. Which brings us back to the question: Why not JSON?

File: src/apple/alohalytics_objc.mm [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+  }
+  std::string preferredLocalizations;
+  for (NSString * loc in [[NSBundle mainBundle] preferredLocalizations]) {
+    preferredLocalizations += [loc UTF8String] + std::string(" ");
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: High-level comment: You seem to do a lot of string packing here. Which brings us back to the question: Why not JSON?

File: src/apple/alohalytics_objc.mm [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+
+// Returns path to store statistics files.
+static std::string StoragePath() {
+  // Store files in special directory which is not backed up automatically.
+  NSArray * paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
+  NSString * directory = [[paths firstObject] stringByAppendingString:@"/Alohalytics/"];
+  NSFileManager * fm = [NSFileManager defaultManager];
+  if (![fm fileExistsAtPath:directory]) {
+    if (![fm createDirectoryAtPath:directory withIntermediateDirectories:YES attributes:nil error:nil]) {
+      NSLog(@"Alohalytics ERROR: Can't create directory %@.", directory);
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: This is a critical error, isn't it? Do I read it right that you proceed with the initialization even if the directory could not have been created?

File: src/apple/alohalytics_objc.mm [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    }
+#if (TARGET_OS_IPHONE > 0)
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: And on an iPad? A comment would be nice here.

File: src/apple/alohalytics_objc.mm [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+    // Disable iCloud backup for storage folder: https://developer.apple.com/library/iOS/qa/qa1719/_index.html
+    const std::string storagePath = [directory UTF8String];
+    if (NSFoundationVersionNumber >= NSFoundationVersionNumber_iOS_5_1) {
+      CFURLRef url = CFURLCreateFromFileSystemRepresentation(kCFAllocatorDefault,
+                                                             reinterpret_cast<unsigned char const *>(storagePath.c_str()),
+                                                             storagePath.size(),
+                                                             0);
+      CFErrorRef err;
+      signed char valueRaw = 1; // BOOL YES
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: `signed char valueOfYes = 1;` ?

File: src/apple/alohalytics_objc.mm [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+      CFNumberRef value = CFNumberCreate(kCFAllocatorDefault, kCFNumberCharType, &valueRaw);
+      if (!CFURLSetResourcePropertyForKey(url, kCFURLIsExcludedFromBackupKey, value, &err)) {
+        NSLog(@"Alohalytics ERROR while disabling iCloud backup for directory %@", directory);
+      }
+      CFRelease(value);
+      CFRelease(url);
+    } else {
+      static char const * attrName = "com.apple.MobileBackup";
+      u_int8_t attrValue = 1;
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: Same here.

File: src/apple/alohalytics_objc.mm [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+  if (source) {
+    parsed.emplace("UIApplicationLaunchOptionsSourceApplicationKey", ToStdString(source));
+  }
+  return parsed;
+}
+} // namespace
+
+@implementation Alohalytics
+
++ (void)setDebugMode:(BOOL)enable {
```
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: I wonder if the indentation here and above is intended.

**Comment by dkorolev** [comment]
LGTM++

File: examples/ios/SampleClient/AppDelegate.m (hunk 2 above)
Comment (deathbaba) [inline examples/ios/SampleClient/AppDelegate.m]: Thanks for pointing to it. It explains everything.

File: src/apple/alohalytics_objc.mm (hunk 5 above)
Comment (deathbaba) [inline src/apple/alohalytics_objc.mm]: - It is still not too-much and local for iOS only
- JSON packing have much more overhead and limitations to values
- It's compatible with already existing statistics solutions
- It does not depend on external json lib in the case we refactor cereal out

File: src/apple/alohalytics_objc.mm (hunk 6 above)
Comment (deathbaba) [inline src/apple/alohalytics_objc.mm]: It's a local case for some specific iOS device's data (not a general case).

File: src/apple/alohalytics_objc.mm (hunk 8 above)
Comment (deathbaba) [inline src/apple/alohalytics_objc.mm]: Right. Even if you can't get any statistics about the product, it should remain usable for the user.
This is the main reason why I prefer to silence all exceptions and critical errors in the statistics core.

File: src/apple/alohalytics_objc.mm (hunk 12 above)
Comment (deathbaba) [inline src/apple/alohalytics_objc.mm]: Yes, it's ObjC-style indent.

**Comment by deathbaba** [comment]
@dkorolev PTAL & merge if it's ok.

**Comment by dkorolev** [comment]
LGTM++

I left a few comments. Looks good to merge anyway, and I'm doing so.

File: src/apple/alohalytics_objc.mm (hunk 12 above)
Comment (dkorolev) [inline src/apple/alohalytics_objc.mm]: Pichalka.

**Pull request merged**